src/lib/profiles/security/WeaveCertProvisioning.cpp

```
@@ -0,0 +1,767 @@
+
+    VerifyOrDie(client->mState == kState_RequestInProgress);
+    VerifyOrDie(client->mEC == ec);
+
+    // Process status report message from the CA service.
+    if (profileId == kWeaveProfile_Common && msgType == Common::kMsgType_StatusReport)
+    {
+        StatusReport statusReport;
+        WEAVE_ERROR parseErr = StatusReport::parse(msgBuf, statusReport);
+
+        // Free the received message buffer so that it can be reused to initiate additional communication.
+        PacketBuffer::Free(msgBuf);
+        msgBuf = NULL;
+
+        SuccessOrExit(parseErr);
+
+        VerifyOrExit(statusReport.mProfileId == kWeaveProfile_Security, /* no-op */);
+
+        InEventParam inParam;
+        OutEventParam outParam;
+        inParam.Clear();
+        outParam.Clear();
+        inParam.Source = client;
+
+        if (statusReport.mStatusCode == Security::kStatusCode_NoNewCertRequired)
+        {
+            inParam.ResponseReceived.ReplaceCert = false;
+
+            // Deliver a ResponseReceived API event to the application.
+            client->mEventCallback(client->AppState, kEvent_ResponseReceived, inParam, outParam);
+        }
+        else
+        {
+            inParam.CommunicationError.Reason = WEAVE_ERROR_STATUS_REPORT_RECEIVED;
+            inParam.CommunicationError.RcvdStatusReport = &statusReport;
+
+            // Deliver a CommunicationError API event to the application.
+            client->mEventCallback(client->AppState, kEvent_CommunicationError, inParam, outParam);
+        }
+    }
+
+    // Process GetCertificateResponse message from the CA service.
+    else if (profileId == kWeaveProfile_Security && msgType == kMsgType_GetCertificateResponse)
+    {
+        // If the request is now complete, clear the request state and enter Idle state.
+        client->HandleRequestDone();
+
+        // Process message and deliver result to the application using API event callback function.
+        client->ProcessGetCertificateResponse(msgBuf);
```

Maybe we should add a comment explaining why we ignore the error return error.  Something like:

_Note that, in this context, any error returned by ProcessGetCertificateResponse() has already been seen by the application.  Thus we can safely ignore it here._

---

done